File: src/modules/landing_target_estimator/KF_xyzb_coupled_moving.cpp
/****************************************************************************
 *
 *   Copyright (c) 2013-2018 PX4 Development Team. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 * 3. Neither the name PX4 nor the names of its contributors may be
 *    used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 ****************************************************************************/

/**
 * @file KF_xyzb_coupled_moving.cpp
 * @brief Filter to estimate the pose of moving targets. State: [rx, ry, rz, r_dotx, r_doty, r_dotz, bx, by, bz, atx, aty, atz]
 *
 * @author Jonas Perolini <jonas.perolini@epfl.ch>
 *
 */

#include "KF_xyzb_coupled_moving.h"
#include "python_derivation/generated/coupled_moving_xyzb/predictCov.h"
#include "python_derivation/generated/coupled_moving_xyzb/computeInnovCov.h"
#include "python_derivation/generated/coupled_moving_xyzb/syncState.h"

namespace landing_target_estimator
{

void KF_xyzb_coupled_moving::predictState(float dt, matrix::Vector<float, 3> acc)
{

	// [rx, ry, rz, r_dotx, r_doty, r_dotz, bx, by, bz, atx, aty, atz]
	const float tmp0 = 0.5f * dt * dt;

	_state(0, 0) = _state(0, 0) + _state(3, 0) * dt + _state(9, 0) * tmp0 - tmp0 * acc(0);
	_state(1, 0) = _state(1, 0) + _state(10, 0) * tmp0 + _state(4, 0) * dt - tmp0 * acc(1);
	_state(2, 0) = _state(11, 0) * tmp0 + _state(2, 0) + _state(5, 0) * dt - tmp0 * acc(2);
	_state(3, 0) = _state(3, 0) + _state(9, 0) * dt - acc(0) * dt;
	_state(4, 0) = _state(10, 0) * dt + _state(4, 0) - acc(1) * dt;
	_state(5, 0) = _state(11, 0) * dt + _state(5, 0) - acc(2) * dt;
}

void KF_xyzb_coupled_moving::predictCov(float dt)
{
	matrix::Matrix<float, 12, 12> cov_updated;
	sym::Predictcov(dt, _input_var, _bias_var, _acc_var, _covariance, &cov_updated);
	_covariance = cov_updated;
}


bool KF_xyzb_coupled_moving::update()
{
	// Avoid zero-division
	if ((float)abs(_innov_cov) < (float)1e-6) {
		return false;
	}

	const float beta = _innov / _innov_cov * _innov;

	// 5% false alarm probability
	if (beta > 3.84f) {
		return false;
	}

	const matrix::Matrix<float, 12, 1> kalmanGain = _covariance * _meas_matrix.transpose() / _innov_cov;

	_state = _state + kalmanGain * _innov;
	_covariance = _covariance - kalmanGain * _meas_matrix * _covariance;

	return true;
}

void KF_xyzb_coupled_moving::setH(matrix::Vector<float, 15> h_meas)
{
	// h_meas = [rx, ry, rz, r_dotx, r_doty, r_dotz, bx, by, bz, atx, aty, atz, vtx, vty, vtz]
	// idx    = [0,   1,  2,      3,      4,      5,  6,  7,  8,   9,  10,  11,  12,  13,  14]
	_meas_matrix(0, 0) = h_meas(0);
	_meas_matrix(0, 1) = h_meas(1);
	_meas_matrix(0, 2) = h_meas(2);
	_meas_matrix(0, 3) = h_meas(3);
	_meas_matrix(0, 4) = h_meas(4);
	_meas_matrix(0, 5) = h_meas(5);
	_meas_matrix(0, 6) = h_meas(6);
	_meas_matrix(0, 7) = h_meas(7);
	_meas_matrix(0, 8) = h_meas(8);
	_meas_matrix(0, 9) = h_meas(9);
	_meas_matrix(0, 10) = h_meas(10);
	_meas_matrix(0, 11) = h_meas(11);
}

void KF_xyzb_coupled_moving::syncState(float dt, matrix::Vector<float, 3> acc)
{
	matrix::Matrix<float, 12, 1> sync_stat_updated;
	sym::Syncstate(dt, _state, acc, &sync_stat_updated);
	_sync_state = sync_stat_updated;
}

float KF_xyzb_coupled_moving::computeInnovCov(float meas_unc)
{
	float innov_cov_updated;
	sym::Computeinnovcov(meas_unc, _covariance, _meas_matrix, &innov_cov_updated);
	_innov_cov = innov_cov_updated;

	return _innov_cov;
}

float KF_xyzb_coupled_moving::computeInnov(float meas)
{
	/* z - H*x */
	_innov = meas - (_meas_matrix * _sync_state)(0, 0);
	return _innov;
}

} // namespace landing_target_estimator

File: src/modules/landing_target_estimator/python_derivation/KF_xyzb_coupled_moving.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    Copyright (c) 2022 PX4 Development Team
    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

    1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
    2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in
    the documentation and/or other materials provided with the
    distribution.
    3. Neither the name PX4 nor the names of its contributors may be
    used to endorse or promote products derived from this software
    without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
    OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
    AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.

File: KF_xyzb_coupled_moving.py
Description: Derive the Kalman filter equations for moving targets with coupled dynamics using symforce
Author: Jonas Perolini <jonas.perolini@epfl.ch>
"""

import symforce.symbolic as sf

def generate_px4_function(function_name, output_names):
    from symforce.codegen import Codegen, CppConfig
    import os
    import fileinput

    codegen = Codegen.function(
            function_name,
            output_names=output_names,
            config=CppConfig())
    metadata = codegen.generate_function(
            output_dir="src/modules/landing_target_estimator/python_derivation/generated/coupled_moving_xyzb",
            skip_directory_nesting=True)

    print("Files generated in {}:\n".format(metadata.output_dir))
    for f in metadata.generated_files:
        print("  |- {}".format(os.path.relpath(f, metadata.output_dir)))

    # Replace cstdlib and Eigen functions by PX4 equivalents
    with fileinput.FileInput(os.path.abspath(metadata.generated_files[0]), inplace=True) as file:
        for line in file:
            line = line.replace("std::max", "math::max")
            line = line.replace("std::min", "math::min")
            line = line.replace("Eigen", "matrix")
            line = line.replace("matrix/Dense", "matrix/math.hpp")
            print(line, end='')

#-------------------------------- COUPLED DYNAMICS, STATIC TARGET ------------------------------------------ #

class State:
    rx = 0
    ry = 1
    rz = 2
    rx_dot = 3
    ry_dot = 4
    rz_dot = 5
    bx = 6
    by = 7
    bz = 8
    atx = 9,
    aty = 10,
    atz = 11,
    n_states = 12

class Input:
    ax = 0
    ay = 1
    az = 2
    n_inputs = 3

class Directions:
    x = 0,
    y = 1,
    z = 2,
    nb_directions = 3

class VState(sf.Matrix):
    SHAPE = (State.n_states, 1)

class MState(sf.Matrix):
    SHAPE = (State.n_states, State.n_states)

class VInput(sf.Matrix):
    SHAPE = (Input.n_inputs, 1)

class VMeas(sf.Matrix):
    SHAPE = (1, State.n_states)

class MDirections(sf.Matrix):
    SHAPE = (Directions.nb_directions, Directions.nb_directions)


def predictState(dt: sf.Scalar, state: VState, acc: VInput) -> (VState):

    identity = sf.Matrix([  [1, 0, 0],
                            [0, 1, 0],
                            [0, 0, 1]])

    Phi = sf.Matrix.block_matrix([   [1 * identity, dt * identity, 0 * identity, 0.5*dt*dt * identity],
                        [0 * identity, 1 * identity,  0 * identity, dt * identity       ],
                        [0 * identity, 0 * identity,  1 * identity, 0  * identity       ],
                        [0 * identity, 0 * identity,  0 * identity, 1  * identity       ]])


    G = sf.Matrix.block_matrix([ [-0.5*dt*dt * identity], [-dt * identity], [0 * identity], [0 * identity]])

    return (Phi * state + G*acc).simplify()

def syncState(dt: sf.Scalar, state: VState, acc: VInput) -> (VState):

    identity = sf.Matrix([  [1, 0, 0],
                            [0, 1, 0],
                            [0, 0, 1]])

    Phi = sf.Matrix.block_matrix([   [1 * identity, dt * identity, 0 * identity, 0.5*dt*dt * identity],
                        [0 * identity, 1 * identity,  0 * identity, dt * identity       ],
                        [0 * identity, 0 * identity,  1 * identity, 0  * identity       ],
                        [0 * identity, 0 * identity,  0 * identity, 1  * identity       ]])


    G = sf.Matrix.block_matrix([ [-0.5*dt*dt * identity], [-dt * identity], [0 * identity], [0 * identity]])

    return (Phi.inv() * (state - G*acc)).simplify()

def predictCov(dt: sf.Scalar, input_var: MDirections, bias_var: MDirections, acc_var: MDirections, covariance: MState) -> (MState):
    identity = sf.Matrix([  [1, 0, 0],
                            [0, 1, 0],
                            [0, 0, 1]])

    Phi = sf.Matrix.block_matrix([   [1 * identity, dt * identity, 0 * identity, 0.5*dt*dt * identity],
                        [0 * identity, 1 * identity,  0 * identity, dt * identity       ],
                        [0 * identity, 0 * identity,  1 * identity, 0  * identity       ],
                        [0 * identity, 0 * identity,  0 * identity, 1  * identity       ]])

    G = sf.Matrix.block_matrix([ [-0.5*dt*dt * identity], [-dt * identity], [0 * identity], [0 * identity]])

    Q_acc_diag = sf.Matrix([  [acc_var[0,0], 0, 0],
                        [0, acc_var[1,1], 0],
                        [0, 0, acc_var[2,2]]])

    Q_acc = sf.Matrix.block_matrix([   [0 * identity, 0 * identity, 0 * identity, 0 * identity],
                    [0 * identity, 0 * identity,  0 * identity, 0 * identity       ],
                    [0 * identity, 0 * identity,  0 * identity, 0  * identity       ],
                    [0 * identity, 0 * identity,  0 * identity,  Q_acc_diag * identity       ]])

    Q_bias_diag = sf.Matrix([  [bias_var[0,0], 0, 0],
                            [0, bias_var[1,1], 0],
                            [0, 0, bias_var[2,2]]])

    Q_bias = sf.Matrix.block_matrix([   [0 * identity, 0 * identity, 0 * identity, 0 * identity],
                    [0 * identity, 0 * identity,  0 * identity, 0 * identity       ],
                    [0 * identity, 0 * identity,  Q_bias_diag, 0  * identity       ],
                    [0 * identity, 0 * identity,  0 * identity, 0  * identity       ]])

    return G*input_var*G.T + Q_bias + Q_acc + Phi*covariance*Phi.T


def computeInnovCov(meas_unc: sf.Scalar, covariance: MState, meas_matrix: VMeas) -> (sf.Scalar):
    return (meas_matrix*covariance*meas_matrix.T)[0,0] + meas_unc

generate_px4_function(predictState, output_names=["state_pred"])
generate_px4_function(syncState, output_names=["state_updated"])
generate_px4_function(predictCov, output_names=["cov_updated"])
generate_px4_function(computeInnovCov, output_names=["innov_cov_updated"])